spatialos-sdk/src/worker/component.rs

```diff
@@ -1,35 +1,463 @@
+                    component_id: update.component_id,
+                    internal: update.schema_type,
+                },
+                user_handle: update.user_handle,
+            }
+        }
+    }
+
+    #[derive(Debug)]
+    pub struct CommandRequest {
+        pub component_id: ComponentId,
+        pub schema_type: SchemaCommandRequest,
+        pub user_handle: *mut Worker_CommandRequestHandle,
+    }
+
+    impl From<&Worker_CommandRequest> for CommandRequest {
+        fn from(request: &Worker_CommandRequest) -> Self {
+            CommandRequest {
+                component_id: request.component_id,
+                schema_type: SchemaCommandRequest {
+                    component_id: request.component_id,
+                    internal: request.schema_type,
+                },
+                user_handle: request.user_handle,
+            }
+        }
+    }
+
+    #[derive(Debug)]
+    pub struct CommandResponse {
+        pub component_id: ComponentId,
+        pub schema_type: SchemaCommandResponse,
+        pub user_handle: *mut Worker_CommandResponseHandle,
+    }
+
+    impl From<&Worker_CommandResponse> for CommandResponse {
+        fn from(response: &Worker_CommandResponse) -> Self {
+            CommandResponse {
+                component_id: response.component_id,
+                schema_type: SchemaCommandResponse {
+                    component_id: response.component_id,
+                    internal: response.schema_type,
+                },
+                user_handle: response.user_handle,
+            }
         }
     }
 }
 
-// TODO: Wrap Schema_ComponentUpdate
-pub struct ComponentUpdate {
-    pub component_id: ComponentId,
-    pub schema_type: *mut Schema_ComponentUpdate,
+// A data structure which represents all known component types. Used to generate an array of vtables to pass
+// to the connection object.
+pub struct ComponentDatabase {
+    component_vtables: Vec<worker::Worker_ComponentVtable>,
+}
+
+impl ComponentDatabase {
+    pub fn new() -> Self {
+        ComponentDatabase {
+            component_vtables: Vec::new(),
+        }
+    }
+
+    pub fn add_component<C: Component>(mut self) -> Self {
+        self.component_vtables.push(create_component_vtable::<C>());
+        self
+    }
+
+    pub(crate) fn to_worker_sdk(&self) -> *const worker::Worker_ComponentVtable {
+        self.component_vtables.as_ptr()
+    }
+
+    pub(crate) fn len(&self) -> usize {
+        self.component_vtables.len()
+    }
+}
+
+// An object which contains a reference counted instance to T.
+struct ClientHandle<T> {
+    data: Rc<T>,
+    // When Drop() is called, data should reduce its reference count.
 }
 
-impl From<&Worker_ComponentUpdate> for ComponentUpdate {
-    fn from(update: &Worker_ComponentUpdate) -> Self {
-        ComponentUpdate {
-            component_id: update.component_id,
-            schema_type: update.schema_type,
+impl<T> ClientHandle<T> {
+    fn new(data: T) -> ClientHandle<T> {
```

Comment:
Honestly, I don't even think you need `ClientHandle<T>` or a `new` function. You could just have `handle_allocate`, `handle_copy`, and `handle_free` as freestanding functions that manipulate an `Arc<T>` under the hood. At least, it doesn't seem to me like there's any value in exposing a `ClientHandle` type to the user when all you're going to do with it is cast it to a `*mut c_void` anyway.

Reply:
Yeah. It didn't feel right to have this extra layer of boxing/indirection, and I felt that it was actually redundant. However, I hadn't had the chance to think about it better after the initial prototyping.